handshake/crypto_setup_client.go

```diff
@@ -83,58 +85,73 @@ func NewCryptoSetupClient(
 		nullAEAD:             crypto.NewNullAEAD(protocol.PerspectiveClient, version),
 		aeadChanged:          aeadChanged,
 		negotiatedVersions:   negotiatedVersions,
+		divNonceChan:         make(chan []byte),
 	}, nil
 }
 
 func (h *cryptoSetupClient) HandleCryptoStream() error {
+	messageChan := make(chan HandshakeMessage)
+	errorChan := make(chan error)
+
+	go func() {
+		for {
+			message, err := ParseHandshakeMessage(h.cryptoStream)
+			if err != nil {
+				errorChan <- qerr.Error(qerr.HandshakeFailed, err.Error())
+				return
+			}
+			messageChan <- message
+		}
+	}()
+
 	for {
 		err := h.maybeUpgradeCrypto()
 		if err != nil {
 			return err
 		}
 
-		// send CHLOs until the forward secure encryption is established
 		h.mutex.RLock()
-		sendCHLO := h.forwardSecureAEAD == nil
+		sendCHLO := h.secureAEAD == nil
 		h.mutex.RUnlock()
+
 		if sendCHLO {
 			err = h.sendCHLO()
 			if err != nil {
 				return err
 			}
 		}
 
-		var shloData bytes.Buffer
-
-		messageTag, cryptoData, err := ParseHandshakeMessage(io.TeeReader(h.cryptoStream, &shloData))
-		if err != nil {
-			return qerr.HandshakeFailed
+		var message HandshakeMessage
+		select {
+		case divNonce := <-h.divNonceChan:
+			if len(h.diversificationNonce) != 0 && !bytes.Equal(h.diversificationNonce, divNonce) {
+				return errConflictingDiversificationNonces
+			}
+			h.diversificationNonce = divNonce
+			// there's no message to process, but we should try upgrading the crypto again
+			continue
+		case msg := <-messageChan:
```

Comment:
This variable is unnecessary, you should be able to read directly into `message`.